restli-client/src/main/java/com/linkedin/restli/client/ExecutionGroup.java

```diff
@@ -0,0 +1,167 @@
+ * Example:
+ * <blockquote>
+ *   <pre>
+ *     ExecutionGroup eg;
+ *     {@code <YourClient>}.get({@code <Parameter>}, eg); // This request will be passed into ExecutionGroup
+ *     eg.execute();
+ *   </pre>
+ * </blockquote>
+ *
+ * (2) Use it inside a lambda function. Corresponding FluentAPIs used inside this lambda will be batched.
+ * Note in this style, you can still optionally pass the Client type as parameter to specifies requests from which
+ * clients the requests need to be batched on. If clients not provided as arguments,
+ * all FluentAPI requests will be batched.
+ *
+ * Example:
+ * <blockquote>
+ *   <pre>
+ *     new ExecutionGroup().batchOn(() -> {
+ *       {@code <YourClient>}.get({@code <Parameter1>});
+ *       {@code <YourClient>}.get({@code <Parameter2>});
+ *     });
+ *   </pre>
+ * </blockquote>
+ * Note: You can use nested executiongroup and each lambda clause have a separate scope.
+ * Example:
+ * <blockquote>
+ *   <pre>
+ *     ExecutionGroup otherEg;
+ *     new ExecutionGroup().batchOn(() -> {
+ *       {@code <YourClient>}.get({@code <Parameter1>}); // implicitly add to the ExecutionGroup which created this lambda
+ *       {@code <YourClient>}.get({@code <Parameter2>}); // implicitly add to the ExecutionGroup which created this lambda
+ *       new ExecutionGroup().batchOn(() -> {
+ *         {@code <YourClient2>}.get({@code <Parameter3>});
+ *         {@code <YourClient2>}.get({@code <Parameter4>});
+ *       }); // this execution group will not be affecting the outer execution group.
+ *       // adding to eg3 explicitly so will not be batched together with other implicit calls in this lambda clause.
+ *       {@code <YourClient>}.get({@code <Parameter3>}, eg3);
+ *     }); // get call from {@code <YourClient>} with parameter1 and parameter2 will be batched
+ *   </pre>
+ * </blockquote>
+ *
+ *
+ */
+@SuppressWarnings({"rawtypes", "unchecked"})
+public class ExecutionGroup
+{
+  private final Map<FluentClient, List<Task>> _clientToTaskListMap = new HashMap<>();
+  private final Engine _engine;
+
+  private Set<FluentClient> _fluentClientAll; // filled by UClient when executionGroup is created; Used for batchOn
+
+  /**
+   * This constructor will be called by the UniversalClient and will not be called by API users directly
+   * @param engine
+   */
+  /* package private */ ExecutionGroup(Engine engine)
+  {
+    _engine = engine;
+  }
+
+  /**
+   * Execute all the tasks that added to {@link ExecutionGroup} through ParSeq Engine
+   */
+  public void execute()
+  {
+    for (Map.Entry<FluentClient, List<Task>> entry : _clientToTaskListMap.entrySet()) {
+      List<Task> taskList = entry.getValue();
+      // the Task.par(Iterable) version does not fast-fail comparing to Task.par(Task...)
+      ParTask<Object> perFluentClientTasks =
+          Task.par(taskList.stream().map(obj -> ((Task<Object>) obj)).collect(Collectors.toList()));
```

Comment:
Can't you make the map store List<Task<?>> so you can avoid this conversion? 
Java stream apis are less performant. It is better to avoid them.

Reply:
Good advice, taken.